### ISSUE TITLE
Issues with overriding functions

### ISSUE THREAD

**Issue by jpziegler** [GaloisInc/saw-script#245]
This issue refers to the following C file:
[ref.c](https://gist.github.com/jpziegler/17f14e055eb6ebd5c21ad2b92f326f41)
And accompanying SAW file:
[ref.saw](https://gist.github.com/jpziegler/b71ace405edd550bfc6eca3ccc0a19d2)

The function [fun](https://gist.github.com/jpziegler/17f14e055eb6ebd5c21ad2b92f326f41#file-ref-c-L21) demonstrates the use of a function with multiple "types" for the called [add](https://gist.github.com/jpziegler/17f14e055eb6ebd5c21ad2b92f326f41#file-ref-c-L1) function. In this case, "types" refers to the fact that the same reference is being passed in for multiple parameters at times.

The [topdown](https://gist.github.com/jpziegler/b71ace405edd550bfc6eca3ccc0a19d2#file-ref-saw-L73) verification (with no overrides) works fine:
```
fun_result <- crucible_llvm_verify m "fun" [] false fun_spec abc;
```
The [naïve](https://gist.github.com/jpziegler/b71ace405edd550bfc6eca3ccc0a19d2#file-ref-saw-L84) compositional verification (with a single override for `add`) does not work. It fails with `No applicable override for add`. In a nutshell, this is because SAW requires that different allocations cannot be the same. This is the first issue; in my opinion, it would be better if we could write a single spec for a function and just use that, though I'm unsure about the details.
```
add_result <- crucible_llvm_verify m "add" [] false add_spec abc;
fun_result <- crucible_llvm_verify m "fun" [add_result] false fun_spec abc;
```
The solution is supposed to be that you write a separate spec for each case and SAW will match with the correct one. This works sometimes, but in this example, it is broken. Here is the [multiple override](https://gist.github.com/jpziegler/b71ace405edd550bfc6eca3ccc0a19d2#file-ref-saw-L88) verification. It fails with `BadPointerLoad`. This is the second issue and is, I think, a genuine bug in SAW.
```
add_result_aaa <- crucible_llvm_verify m "add" [] false add_spec_aaa abc;
add_result_aab <- crucible_llvm_verify m "add" [] false add_spec_aab abc;
add_result_aba <- crucible_llvm_verify m "add" [] false add_spec_aba abc;
add_result_baa <- crucible_llvm_verify m "add" [] false add_spec_baa abc;
add_result_cba <- crucible_llvm_verify m "add" [] false add_spec_cba abc;
fun_result <- crucible_llvm_verify m "fun" [add_result_aaa, add_result_aab, add_result_aba, add_result_baa, add_result_cba] false fun_spec abc;
```
The only workaround I've found is to edit the C source and create a version where each call to `add` is actually a separate function. Then you can have a single override for each function and the problem is solved. Here is the [multiple single-override](https://gist.github.com/jpziegler/b71ace405edd550bfc6eca3ccc0a19d2#file-ref-saw-L76) solution:
```
add_result_aaa <- crucible_llvm_verify m "add_aaa" [] false add_spec_aaa abc;
add_result_aab <- crucible_llvm_verify m "add_aab" [] false add_spec_aab abc;
add_result_aba <- crucible_llvm_verify m "add_aba" [] false add_spec_aba abc;
add_result_baa <- crucible_llvm_verify m "add_baa" [] false add_spec_baa abc;
add_result_cba <- crucible_llvm_verify m "add_cba" [] false add_spec_cba abc;
ezfun_result <- crucible_llvm_verify m "ezfun" [add_result_aaa, add_result_aab, add_result_aba, add_result_baa, add_result_cba] false fun_spec abc;
```

**Comment by atomb** [comment]
With the latest version of SAW, both the multi-function and single-function multi-override cases work.

The single override case, as you describe, does not work. At the moment, this is by design, though there certainly is a convenience argument for having a single specification. My inclination is to focus on ways to make it easy to share the commonality between multiple specifications in cases like this, rather than merging multiple aliasing configurations into one specification.

In `java_verify`, we do provide the `java_may_alias` command to allow a variety of aliasing configurations in one specification block. However, this makes it impossible (with the current symbolic execution approach, at least) to use those specifications to simply extract models, without verification, and then do analysis of those models as a separate pass. The advantages of separating modeling from analysis seem compelling to me, and I expect that we can find a way to reduce repetition in specifications while still allowing this phase separation.

**Comment by brianhuffman** [comment]
Here's an idea: In addition to `crucible_alloc`, we could introduce a new `crucible_alloc_readonly` command for function specs. When verifying the override, those memory regions would be marked as read-only in the symbolic simulator (a feature that doesn't yet exist in crucible but should). When executing the override, we would still assert disjointness of all `crucible_alloc` memory regions, but `crucible_alloc_readonly` allocations wouldn't have to be disjoint from each other. This way we could get by with just one spec for each function in many cases, and I think extraction would still work.

One complication is how to deal with pointer comparisons during verification of an override. E.g. a function might have an explicit aliasing check and do something different in each case. We need to ensure that the result of such a comparison is either a symbolic value or an error, so that the override is also valid in the aliased case.

**Comment by jpziegler** [comment]
@atomb Excellent! We'll try this out on our code and close this ticket if everything works. We can open a new ticket if we really want to bang on override case reduction.

@brianhuffman Could all regions be treated as read-only until a write is observed? That way, we get the behavior you describe without having to add the complexity of using `crucible_alloc_readonly`.

In any case, it sounds like this won't help us when we have a memory region that is passed as both an "input" and "output" parameter. Still, it would be an improvement!

**Comment by brianhuffman** [comment]
Ok, `crucible_alloc_readonly` is now implemented as of a68f3ee771c86bf5101ade7402556298f09511e0.
